### PR TITLE
ARROW-3584: [Go] Implement Table, Schema and Column

### PR DESCRIPTION
**Pull request by sbinet**
This CL adds the `Table` interface and a default simple, non-lazy implementation for it, based off a `Column` and `Schema`.

The `Schema` type is migrated from the `arrow/metadata` package to the main `arrow` one.
Also, `KeyValueMetadata` has been renamed to simply `Metadata`.

### CODE REVIEW

**Comment by sbinet** [comment]
PTAL @stuartcarnie 

especially the memory ownership / ref-count model.
it seems to me the `Retain/Release` strategy starts to be a tad unwieldly with `Columns`, sub-slice of `Columns` and `Table`s.

was `Retain/Release` introduced b/c of possible `mmap`-backed memory allocators?
couldn't we replace those with just a main `Closer interface {Close() error}` interface?

File: go/arrow/array/table.go [hunk 1]
```diff
@@ -126,3 +195,61 @@ func (a *Chunked) NewSlice(i, j int64) *Chunked {
...
+	tbl.validate()
+
+	return &tbl
+}
+
+func (tbl *simpleTable) Schema() *arrow.Schema { return tbl.schema }
+func (tbl *simpleTable) NumRows() int          { return tbl.rows }
+func (tbl *simpleTable) NumCols() int          { return len(tbl.cols) }
+func (tbl *simpleTable) Column(i int) *Column  { return &tbl.cols[i] }
+
```
Comment (stuartcarnie) [inline go/arrow/array/table.go]: Similar to the `Chunked` implementation, `Table` should have a `Retain` and `Release` methods

```go
// Retain increases the reference count by 1.
// Retain may be called simultaneously from multiple goroutines.
func (tbl *simpleTable) Retain() {
	atomic.AddInt64(&tbl.refCount, 1)
}

// Release decreases the reference count by 1.
// When the reference count goes to zero, the memory is freed.
// Release may be called simultaneously from multiple goroutines.
func (a *Chunked) Release() {
	debug.Assert(atomic.LoadInt64(&tbl.refCount) > 0, "too many releases")

	if atomic.AddInt64(&tbl.refCount, -1) == 0 {
		for i := range a.cols {
			cols[i].Release()
		}
		a.chunks = nil
		a.length = 0
		a.nulls = 0
	}
}
```

**Review by stuartcarnie — CHANGES_REQUESTED** [review]
I've added comments about adding reference counting semantics to the `Table` type, which should simplify dealing with the columns.

The `Retain` / `Release` model is expected to integrate with a pooling memory allocator in the future, allowing allocated blocks to return to a pool for reuse. The intension is to reduce GC pressure. Also, as you suspected, it could be used with mmap'd sources to determine when the memory can be unmapped

File: go/arrow/array/table.go [hunk 2]
```diff
@@ -126,3 +195,61 @@ func (a *Chunked) NewSlice(i, j int64) *Chunked {
...
+	}
+	if tbl.rows < 0 {
+		switch len(tbl.cols) {
+		case 0:
+			tbl.rows = 0
+		default:
+			tbl.rows = tbl.cols[0].Len()
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (sbinet) [reply go/arrow/array/table.go]: done.

File: go/arrow/array/table.go (hunk 1 above)
Comment (sbinet) [reply go/arrow/array/table.go]: done.

**Comment by sbinet** [comment]
PTAL

wrt the pool allocator to reduce GC pressure: ok. but I am wondering whether we couldn't "box" this and stop the Retain/Release at the `array.Data` level... I'll ponder on this some more...
thanks!

**Comment by wesm** [comment]
Going to merge this to help these patches move along, further feedback from @stuartcarnie can be addressed in a follow up patch if needed

**Pull request closed without merging**